### ISSUE TITLE
Cannot build Xamarin.IOS project on Mac -> missing parameterless constructor?

### ISSUE THREAD

**Issue by PaulVrugt** [Microsoft/vsts-agent#1373]
## Agent version and platform

Version of your agent?
2.122.1

OS of the machine running the agent? 
OSX

## VSTS type and version
On-Prem TFS

If On-Prem TFS, which release
2017 Update 3

## What's not working?
I'm setting up a build environment for build our xamarin.forms apps via tfs builds. I've successfully set up the android build on a windows build machine, now I'm trying to build the IOS app on a Mac. I've installed an configured the OSX build agent, and installed visual studio for mac. Now when building our app, I get an error that the Version object (System.Version) doesn't have a parameterless constructor. The error is in the .net Standard 2.0 project which is the shared library used in both versions of the app:

General/Settings.cs(62,32): error CS1729: 'Version' does not contain a constructor that takes 0 arguments 

The compiler used is:
/Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/4.5/csc.exe
might this be the wrong one? and if so. Why it is using this compiler?

I am using msbuild 15 (which is confirmed in the build log). This is full compiler command being executed:
[completecsccommand.txt](https://github.com/Microsoft/vsts-agent/files/1639711/completecsccommand.txt)
The system.Version we use from .net standard 2.0 is in the netstandard.dll of .net standard 2.0. Might it be that a wrong version of .net standard is being used for compiling? And if so. Why? Am I doing something wrong, or is it a bug?

Now this guide: https://docs.microsoft.com/en-us/vsts/build-release/apps/mobile/xamarin?tabs=vsts
states that I should create a separate configuration for IOS builds containing only the IOS app project, but this doesn't make any sense, because then I get all kinds of errors about classes from the shared library missing (duh, the shared library is not built in this configuration). However, I found a different guide (https://blog.xamarin.com/continuous-integration-for-ios-apps-with-visual-studio-team-services/) that DOES enable the shared projects in the IOS configuration. However, that causes the parameterless constructor error to re-appear. 

I might be doing something fundamentally wrong. But I've googled my ass off, and there seems to be no guide in how to do this.

## Agent and Worker's diag log
[Agent_20180117-154648-utc.log](https://github.com/Microsoft/vsts-agent/files/1639651/Agent_20180117-154648-utc.log)

**Comment by TingluoHuang** [comment]
@PaulVrugt can you build your solution in Visual Studio for Mac? I guess you didn't install dotnet 2.x SDK on your Mac. Can you run `dotnet --version`?

**Comment by PaulVrugt** [comment]
Running dotnet --version returns 2.1.4

I cannot build the solution in visual studio for mac, because it starts complaining about a missing packages.config file, which I indeed don't have. I've googled this, and this seems to be a problem with visual studio for mac not understanding projects without a packages.config. However, when having the solution open in visual studio for mac, it does recognize the Version constructor and I can go to the definition, and it opens the netstandard assembly in the assembly browser.

**Comment by TingluoHuang** [comment]
@PaulVrugt can you share a full build log?
https://docs.microsoft.com/en-us/dotnet/api/system.version.-ctor?view=netcore-1.1#System_Version__ctor
you definitely hit a wrong dotnet sdk version, since `New Version()`  is introduced in netstandard 2.0

**Comment by PaulVrugt** [comment]
@TingluoHuang 

I've created the log for 2 builds. One containing a failed build via the build agent that uses the wrong sdk, the other is a successful build of the same project using visual studio for mac. The sources used are identical.

Failed build: 
[FailedBuild.zip](https://github.com/Microsoft/vsts-agent/files/1642019/FailedBuild.zip)

Successful build:
[SuccessfulBuild.zip](https://github.com/Microsoft/vsts-agent/files/1642023/SuccessfulBuild.zip)

I've taken a look at the logfiles myselft, and the successful build clearly uses a different version of .net standard in the references list in the "csc.exe" command. The big question howerver, is why, since the sources used are identical.

**Comment by TingluoHuang** [comment]
@madhurig for Xamarin.IOS task question.

**Comment by madhurig** [comment]
It seems like an issue with the version of "nuget" on your agent. Please try:
1. dotnet restore
2. Nuget restore with latest version (4.5). We have a nuget tool installer task but that is only for vsts: https://docs.microsoft.com/en-us/vsts/build-release/tasks/tool/nuget
3. Run xamarin.ios but choose the option to skip restoring packages

Thanks,
Madhuri

**Comment by PaulVrugt** [comment]
@madhurig 

I think I understand what you mean. And if so, I'm afraid it didn't help. What i did:

1. I performed a dotnet restore on the build agent server of the project failing to build
2. I disabled the nuget restore options in the xamarin.ios build task
3. I added the nuget tool installer to the build definition to install nuget 4.5
4. I tried 2 things:
4a. Manually use the installed nuget 4.5 on the build agent machine to restore the packages and run the build -> same issue, Version constructor error
4b. I added a separate nuget restore task to the build definition and run the build. However, the nuget restore task failed because it doesn't use the 4.5 version of nuget defined by the nuget tool installer task, but it uses nuget 4.0, which (i guess) doesn't understand including .net framework dll's in .net standard 2.0 projects.

Did I do what you meant? What else can we try to fix this?

**Comment by PaulVrugt** [comment]
@madhurig 

I did another try:

- I cleaned all package cache from the agent machine
- performed a dotnet restore
- manually restored all packages of the solution i'm trying to build
- disabled package restore in the xamarin.ios task
- run a build

This causes the build to fail with loads of errors, about the all kinds of basic classes and namespaces not being found like System.String or System.Reflection.

I then enabled the package restore again in the xamairn.ios build. It then started restoring all kinds of packages during the build, and the original error re-appeared. (version has no constructor with 0 arguments). Here are the logs of the last failed build 
[FailedBuild2.zip](https://github.com/Microsoft/vsts-agent/files/1646226/FailedBuild2.zip)
